File: src/bank_api/persistence/repositories.py
"""Repository classes wrapping SQLAlchemy sessions."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Sequence

from sqlalchemy import select
from sqlalchemy.orm import Session

from ..models.banking import AccountBalance, AccountTransaction
from ..models.common import DateString
from .models import Order, Position, Setting, SyncLog, Transaction


class TransactionRepository:
    """Persist and retrieve account transactions."""

    def __init__(self, session: Session) -> None:
        self._session = session

    def upsert_transactions(
        self, transactions: Sequence[AccountTransaction], account_id: str
    ) -> None:
        for transaction in transactions:
            external_id = transaction.reference or transaction.endToEndReference
            existing = None
            if external_id:
                existing = self._session.scalar(
                    select(Transaction).where(Transaction.external_id == external_id)
                )
            amount_value = None
            amount = transaction.amount
            if amount and amount.value is not None:
                amount_value = float(amount.value)
            currency = amount.unit if amount and amount.unit else None
            if existing:
                existing.raw = transaction.model_dump()
                existing.amount = amount_value
                existing.currency = currency
                existing.booking_date = _parse_date(transaction.bookingDate)
            else:
                db_obj = Transaction(
                    external_id=external_id,
                    account_id=account_id,
                    booking_date=_parse_date(transaction.bookingDate),
                    amount=amount_value,
                    currency=currency,
                    raw=transaction.model_dump(),
                )
                self._session.add(db_obj)

    def list_transactions(self, account_id: str) -> list[Transaction]:
        stmt = select(Transaction).where(Transaction.account_id == account_id)
        return list(self._session.scalars(stmt))


class PositionRepository:
    """Persist account balances (positions)."""

    def __init__(self, session: Session) -> None:
        self._session = session

    def upsert_balances(self, balances: Sequence[AccountBalance]) -> None:
        for balance in balances:
            account_id = balance.accountId or balance.account.accountId if balance.account else None
            if not account_id:
                continue
            existing = self._session.scalar(
                select(Position).where(Position.account_id == account_id)
            )
            if existing:
                existing.balance_amount = _extract_amount(balance)
                existing.currency = _extract_currency(balance)
                existing.raw = balance.model_dump()
                existing.updated_at = datetime.utcnow()
            else:
                db_obj = Position(
                    account_id=account_id,
                    balance_amount=_extract_amount(balance),
                    currency=_extract_currency(balance),
                    raw=balance.model_dump(),
                )
                self._session.add(db_obj)

    def list_positions(self) -> list[Position]:
        return list(self._session.scalars(select(Position)))


class SyncLogRepository:
    """Persist logs for background jobs."""

    def __init__(self, session: Session) -> None:
        self._session = session

    def create(self, job_name: str, status: str, detail: str | None = None) -> SyncLog:
        entry = SyncLog(
            job_name=job_name, status=status, detail=detail, started_at=datetime.utcnow()
        )
        self._session.add(entry)
        return entry

    def update(self, log: SyncLog, status: str, detail: str | None = None) -> None:
        log.status = status
        log.detail = detail
        log.finished_at = datetime.utcnow()

    def get(self, log_id: int) -> SyncLog | None:
        return self._session.get(SyncLog, log_id)


class SettingsRepository:
    """Persist simple key/value configuration entries."""

    def __init__(self, session: Session) -> None:
        self._session = session

    def get(self, key: str) -> Setting | None:
        return self._session.scalar(select(Setting).where(Setting.key == key))

    def set(self, key: str, value: str | None) -> Setting:
        setting = self.get(key)
        if setting:
            setting.value = value
        else:
            setting = Setting(key=key, value=value)
            self._session.add(setting)
        return setting

    def list_all(self) -> list[Setting]:
        return list(self._session.scalars(select(Setting)))


class OrderRepository:
    """Manage lifecycle of locally stored orders."""

    def __init__(self, session: Session) -> None:
        self._session = session

    def list_orders(self) -> list[Order]:
        stmt = select(Order).order_by(Order.created_at.desc())
        return list(self._session.scalars(stmt))

    def get(self, order_id: int) -> Order | None:
        return self._session.get(Order, order_id)

    def create(
        self,
        *,
        instrument: str,
        side: str,
        order_type: str,
        quantity: float,
        limit_price: float | None,
        notes: str | None = None,
    ) -> Order:
        order = Order(
            instrument=instrument,
            side=side,
            order_type=order_type,
            quantity=quantity,
            limit_price=limit_price,
            notes=notes,
        )
        self._session.add(order)
        self._session.flush()
        return order

    def update_status(self, order: Order, status: str) -> Order:
        order.status = status
        return order


def _parse_date(date_value: Any) -> datetime | None:
    if not date_value:
        return None
    if isinstance(date_value, DateString):
        return datetime.combine(date_value.date, datetime.min.time())
    try:
        return datetime.fromisoformat(str(date_value))
    except ValueError:
        return None


def _extract_amount(balance: AccountBalance) -> float | None:
    if balance.balance and balance.balance.value is not None:
        return float(balance.balance.value)
    if balance.availableCashAmount and balance.availableCashAmount.value is not None:
        return float(balance.availableCashAmount.value)
    return None


def _extract_currency(balance: AccountBalance) -> str | None:
    if balance.balance and balance.balance.unit:
        return balance.balance.unit
    if balance.availableCashAmount and balance.availableCashAmount.unit:
        return balance.availableCashAmount.unit
    return None

File: node/src/persistence/repositories.js
function createTransactionRepository(db) {
  return {
    upsertTransactions(transactions, accountId) {
      const select = db.prepare("SELECT id FROM transactions WHERE external_id = ?");
      const insert = db.prepare(`
        INSERT INTO transactions (external_id, account_id, booking_date, amount, currency, raw)
        VALUES (@external_id, @account_id, @booking_date, @amount, @currency, @raw)
      `);
      const update = db.prepare(`
        UPDATE transactions
        SET account_id = @account_id,
            booking_date = @booking_date,
            amount = @amount,
            currency = @currency,
            raw = @raw,
            updated_at = CURRENT_TIMESTAMP
        WHERE external_id = @external_id
      `);

      const run = db.transaction((items) => {
        for (const transaction of items || []) {
          const externalId = transaction.reference || transaction.endToEndReference || null;
          const row = mapTransaction(transaction, accountId, externalId);
          if (externalId && select.get(externalId)) {
            update.run(row);
          } else {
            insert.run(row);
          }
        }
      });

      run(transactions);
    },

    listTransactions(accountId) {
      const rows = db
        .prepare("SELECT raw FROM transactions WHERE account_id = ? ORDER BY booking_date DESC, id DESC")
        .all(accountId);
      return rows.map((row) => parseJson(row.raw)).filter(Boolean);
    }
  };
}

function createPositionRepository(db) {
  return {
    upsertBalances(balances) {
      const statement = db.prepare(`
        INSERT INTO positions (account_id, balance_amount, currency, raw, updated_at)
        VALUES (@account_id, @balance_amount, @currency, @raw, CURRENT_TIMESTAMP)
        ON CONFLICT(account_id) DO UPDATE SET
          balance_amount = excluded.balance_amount,
          currency = excluded.currency,
          raw = excluded.raw,
          updated_at = CURRENT_TIMESTAMP
      `);

      const run = db.transaction((items) => {
        for (const balance of items || []) {
          const accountId = balance.accountId || (balance.account && balance.account.accountId);
          if (!accountId) {
            continue;
          }
          statement.run({
            account_id: accountId,
            balance_amount: extractAmount(balance),
            currency: extractCurrency(balance),
            raw: JSON.stringify(balance)
          });
        }
      });

      run(balances);
    },

    listPositions() {
      const rows = db.prepare("SELECT raw FROM positions ORDER BY account_id").all();
      return rows.map((row) => parseJson(row.raw)).filter(Boolean);
    }
  };
}

function createSettingsRepository(db) {
  return {
    get(key) {
      return db.prepare("SELECT key, value, updated_at FROM settings WHERE key = ?").get(key) || null;
    },

    set(key, value) {
      db.prepare(`
        INSERT INTO settings (key, value, updated_at)
        VALUES (?, ?, CURRENT_TIMESTAMP)
        ON CONFLICT(key) DO UPDATE SET
          value = excluded.value,
          updated_at = CURRENT_TIMESTAMP
      `).run(key, value === undefined ? null : value);
      return this.get(key);
    },

    listAll() {
      return db.prepare("SELECT key, value, updated_at FROM settings ORDER BY key").all();
    }
  };
}

function createOrderRepository(db) {
  return {
    listOrders() {
      return db.prepare("SELECT * FROM orders ORDER BY created_at DESC, id DESC").all();
    },

    get(orderId) {
      return db.prepare("SELECT * FROM orders WHERE id = ?").get(orderId) || null;
    },

    create(orderData) {
      const result = db.prepare(`
        INSERT INTO orders (instrument, side, order_type, quantity, limit_price, notes)
        VALUES (?, ?, ?, ?, ?, ?)
      `).run(
        orderData.instrument,
        orderData.side,
        orderData.orderType,
        orderData.quantity,
        orderData.limitPrice === undefined ? null : orderData.limitPrice,
        orderData.notes || null
      );
      return this.get(result.lastInsertRowid);
    },

    updateStatus(orderId, status) {
      db.prepare("UPDATE orders SET status = ? WHERE id = ?").run(status, orderId);
      return this.get(orderId);
    }
  };
}

function createSyncLogRepository(db) {
  return {
    create(jobName, status, detail = null) {
      const result = db.prepare(`
        INSERT INTO sync_logs (job_name, status, detail, started_at)
        VALUES (?, ?, ?, CURRENT_TIMESTAMP)
      `).run(jobName, status, detail);
      return this.get(result.lastInsertRowid);
    },

    update(logId, status, detail = null) {
      db.prepare(`
        UPDATE sync_logs
        SET status = ?, detail = ?, finished_at = CURRENT_TIMESTAMP
        WHERE id = ?
      `).run(status, detail, logId);
      return this.get(logId);
    },

    get(logId) {
      return db.prepare("SELECT * FROM sync_logs WHERE id = ?").get(logId) || null;
    }
  };
}

function mapTransaction(transaction, accountId, externalId) {
  const amount = transaction.amount || {};
  return {
    external_id: externalId,
    account_id: accountId,
    booking_date: transaction.bookingDate || null,
    amount: amount.value === undefined || amount.value === null ? null : Number(amount.value),
    currency: amount.unit || null,
    raw: JSON.stringify(transaction)
  };
}

function extractAmount(balance) {
  if (balance.balance && balance.balance.value !== undefined && balance.balance.value !== null) {
    return Number(balance.balance.value);
  }
  if (
    balance.availableCashAmount &&
    balance.availableCashAmount.value !== undefined &&
    balance.availableCashAmount.value !== null
  ) {
    return Number(balance.availableCashAmount.value);
  }
  return null;
}

function extractCurrency(balance) {
  if (balance.balance && balance.balance.unit) {
    return balance.balance.unit;
  }
  if (balance.availableCashAmount && balance.availableCashAmount.unit) {
    return balance.availableCashAmount.unit;
  }
  return null;
}

function parseJson(value) {
  if (!value) {
    return null;
  }
  try {
    return JSON.parse(value);
  } catch (_error) {
    return null;
  }
}

module.exports = {
  createOrderRepository,
  createPositionRepository,
  createSettingsRepository,
  createSyncLogRepository,
  createTransactionRepository
};
